internal/saucecloud/cucumber.go

```diff
@@ -82,6 +66,31 @@ func (r *CucumberRunner) RunProject() (int, error) {
 	return exitCode, nil
 }
 
+func (r *CucumberRunner) validateFramework(m framework.Metadata) (string, error) {
+	r.Project.Playwright.Version = m.FrameworkVersion
+	if r.Project.RunnerVersion == "" {
+		r.Project.RunnerVersion = m.CloudRunnerVersion
+	}
+
+	var deprecationMessage string
+	if m.IsDeprecated() && !m.IsFlaggedForRemoval() {
+		deprecationMessage = r.deprecationMessage(playwright.Kind, r.Project.Playwright.Version, m.RemovalDate)
+		fmt.Print(deprecationMessage)
+	}
+	if m.IsFlaggedForRemoval() {
+		deprecationMessage = r.flaggedForRemovalMessage(playwright.Kind, r.Project.Playwright.Version)
+		fmt.Print(deprecationMessage)
+	}
+
+	for _, s := range r.Project.Suites {
+		if s.PlatformName != "" && !framework.HasPlatform(m, s.PlatformName) {
+			msg.LogUnsupportedPlatform(s.PlatformName, framework.PlatformNames(m.Platforms))
+			return deprecationMessage, errors.New("unsupported platform")
+		}
+	}
+	return deprecationMessage, nil
```

Comment:
The deprecation message is logged in this method (see above) and also returned just do be logged again by the caller. As a developer reading this code, I now have to follow the trail of where the message ends up being used. If we're already refactoring this logic, wouldn't it be better to keep the concern in one single place? I'd recommend to have this method log the deprecation message and not return it.

Reply:
I agree it's weird to get the deprecation message returned in a validation function. So your suggestion is only to log the deprecation at the beginning of the test and do not log it after the test run?

Reply:
Having it _only_ at the end would have the benefit that it's potentially more visible to the user, especially when it's running in CI. But if that makes the code worse, by being less readable, I'd lean towards having better code in this case and just log it at the beginning.

One thing is certain: I don't think we need to have it both at the beginning and end. That seems more like a workaround than a solution.